s3-source-connector/src/main/java/io/aiven/kafka/connect/s3/source/utils/SourceRecordIterator.java

```diff
@@ -42,190 +42,228 @@
-                    final Map<String, Object> partitionMap) {
-
-                long currentOffset;
-
-                if (offsetManager.getOffsets().containsKey(partitionMap)) {
-                    LOGGER.info("***** offsetManager.getOffsets() ***** {}", offsetManager.getOffsets());
-                    currentOffset = offsetManager.incrementAndUpdateOffsetMap(partitionMap, currentObjectKey,
-                            startOffset);
-                } else {
-                    LOGGER.info("Into else block ...");
-                    currentOffset = startOffset;
-                    offsetManager.createNewOffsetMap(partitionMap, currentObjectKey, currentOffset);
-                }
-
-                final Map<String, Object> offsetMap = offsetManager.getOffsetValueMap(currentObjectKey, currentOffset);
 
-                return new S3SourceRecord(partitionMap, offsetMap, topic, topicPartition, key, value, currentObjectKey);
-            }
-
-            @Override
-            public boolean hasNext() {
-                return internalIterator.hasNext();
-            }
+        final Map<String, Object> offsetMap = offsetManager.getOffsetValueMap(objectKey, currentOffset);
 
-            @Override
-            public S3SourceRecord next() {
-                return internalIterator.next();
-            }
-        };
+        return new S3SourceRecord(partitionMap, offsetMap, fileMatcherFilter.getTopicName(),
+                fileMatcherFilter.getPartitionId(), key, value, objectKey);
     }
 
     @Override
     public boolean hasNext() {
-        return recordIterator.hasNext() || objectListIterator.hasNext();
+        while (!workingIterator.hasNext() && iteratorIterator.hasNext()) {
+            workingIterator = iteratorIterator.next();
+        }
+        return workingIterator.hasNext();
     }
 
     @Override
     public S3SourceRecord next() {
-        if (!recordIterator.hasNext()) {
-            nextS3Object();
-        }
-
-        if (!recordIterator.hasNext()) {
-            // If there are still no records, return null or throw an exception
-            return null; // Or throw new NoSuchElementException();
+        if (!hasNext()) {
+            throw new IllegalArgumentException();
         }
-
-        return recordIterator.next();
+        return workingIterator.next();
     }
 
     @Override
     public void remove() {
         throw new UnsupportedOperationException("This iterator is unmodifiable");
     }
 
+    /**
+     * A filter for S3ObjectSummaries that extracts the topic name and partition id. Package private for testing.
+     */
+    static class FileMatcherFilter implements Predicate<S3ObjectSummary> {
+        private static final String PATTERN_TOPIC_KEY = "topicName";
+        private static final String PATTERN_PARTITION_KEY = "partitionId";
+
+        private static final Pattern FILE_DEFAULT_PATTERN = Pattern.compile("(?<topicName>[^/]+?)-"
```

Comment:
I think we need to create an issue for how this is hardcoded and how we are handling it, not sure if @RyanSkraba has this handled in his work